L58-python-async-api.md

```diff
@@ -0,0 +1,1673 @@
+we are providing the only implementation for the interface. We should
+convert them into concrete classes.
+
+On the other hand, there are actually one valid use case that we should keep
+abstract class -- interceptors. To be more specific, the following interfaces
+won't be replaced by concrete classes:
+
+* `grpc.ServerInterceptor`
+* `grpc.UnaryUnaryClientInterceptor`
+* `grpc.UnaryStreamClientInterceptor`
+* `grpc.StreamUnaryClientInterceptor`
+* `grpc.StreamStreamClientInterceptor`
+
+## Future Features
+
+### Implicit Context Propagation By `contextvars`
+
+`contextvars` is a Python 3.7 feature that allows applications to set coroutine
+local variables (just like thread local variables). In the past, gRPC Python
+uses thread local to achieve two goals:
+
+* Support distributed tracing library that work across languages;
+* Support deadline propagation along the chain of RPCs.
+
+For distributed tracing library (like OpenCensus), the tracing information is
+preserved as one of the metadata, and it is expected to be implicitly promoted
+from inbound metadata to outbound metadata. The tracing information will be used
+to monitor the life cycle of an request, and the length it has stayed in each
+services.
+
+For deadline propagation, the deadline of upstream server will be implicitly
+pass down to downstream server, so downstream services can react to that
+information and save computation resources or perform flow control.
+
+Acceptence critiria of this feature:
+* The implementation of such feature should supported by official package, and
+users are not expected to directly access those metadata;
+* Application logic has higher priority than the implicit propagation (e.g.
+  setting timeout explicitly will override the propagated value);
+* Application has the choice to diverse coroutine local variables;
+* The exception error string should be informative (e.g. pointing out the
+  timeout is due to upstream deadline).
+
+Further discussion around this topic, please see related sections under [Rationale].
+
+### Introduce Typing To Generated Code
+
+Typing makes a difference. This gRFC intended to introduce typing to gRPC Python
+as much as we can. To change generated code, we need to update the gRPC protoc
+plugin to generate Python 3 stubs and servicers. To date, **this gRFC intended to**
+**make the new async API compatible with existing generated code**. So, the Python 3
+generated code will have two options:
+
+1. Add a third generation mode to protoc plugin that generates stubs / servicers
+   with typing;
+2. Generating an additional file for the typed code.
+
+## Rationale
+
+### Compatibility With Other Asynchronous Libraries
+
+By other asynchronous libraries, they mean libraries that provide their own
+Future, Coroutine, Event-loops, like `gevent`/`Twisted`/`Tornado`. In general,
+it is challenging to support multiple async mechanisms inter-operate together,
+and it has to be discussed case by case.
```

Comment:
I would also consider the newer async libraries like [Curio](https://github.com/dabeaz/curio) and [Trio](https://github.com/python-trio/trio). Each has thousands of stars and is an alternative to `asyncio`.